src/Build/BackEnd/Node/OutOfProcServerNode.cs

```diff
@@ -438,112 +438,316 @@ internal sealed class RedirectConsoleWriter : TextWriter
         {
             private readonly Action<string> _writeCallback;
             private readonly Timer _timer;
-            private readonly TextWriter _syncWriter;
-
+            private readonly object _lock = new();
             private readonly StringWriter _internalWriter;
 
             private RedirectConsoleWriter(Action<string> writeCallback)
             {
                 _writeCallback = writeCallback;
                 _internalWriter = new StringWriter();
-                _syncWriter = Synchronized(_internalWriter);
                 _timer = new Timer(TimerCallback, null, 0, 40);
             }
 
             public override Encoding Encoding => _internalWriter.Encoding;
 
-            public static TextWriter Create(Action<string> writeCallback)
+            public override void Flush()
             {
-                RedirectConsoleWriter writer = new RedirectConsoleWriter(writeCallback);
+                lock (_lock)
+                {
+                    var sb = _internalWriter.GetStringBuilder();
+                    string captured = sb.ToString();
+                    sb.Clear();
 
-                return writer;
+                    _writeCallback(captured);
```

Comment:
Does the callback need to be executed locked with this lock?

Reply:
It was before (with `[Synchronized]`) so I left it here. I think it needs to be sorta locked since you want to make sure one `_writeCallback` finishes before the next one starts. I think we could split the lock so that one pipe write happens while the next buffer is filling--think that's worth doing?

Reply:
No wait, just two locks could introduce an ordering where `captured` gets overwritten with new data before the old one gets flushed--I think it needs to be in the lock, barring a fancier strategy.

Reply:
Previously I haven't seen the rest of the code - just a passed-in delegate call in lock - that catched my eye.
Since it's all under our controll - it can just stay as it is


Otherwise we'd need to nest the locks - something like:


```csharp
lock(_flushLock)
{
  lock(_lock)
  {
     // ... getting the captured string and flushing the internal writter
  }
  _writeCallback(captured);
}
```

but since the callback today just enqueues, it's likely unnecessary overoptimisation.

Reply:
That's a good idea but I think I'll leave it--I agree it doesn't seem to be hot enough and if it turns out to be we can change it then.

Reply:
100%